silo-to-silo/contracts/src/SiloToSilo.sol

```diff
@@ -0,0 +1,348 @@
+        string memory nearTokenAccountId = registeredTokens[token].nearTokenAccountId;
+        require(bytes(nearTokenAccountId).length > 0, "The token is not registered");
+
+        PromiseCreateArgs memory callStorageDeposit = near.call(
+            nearTokenAccountId,
+            "storage_deposit",
+            bytes(string.concat('{"account_id": "', recipientId, '", "registration_only": true }')),
+            storageDepositAmount,
+            BASE_NEAR_GAS
+        );
+
+        PromiseCreateArgs memory callback = near.auroraCall(
+            address(this),
+            abi.encodeWithSelector(this.storageDepositCallback.selector, token, recipientId, isRecipientStorageDeposit),
+            NO_DEPOSIT,
+            BASE_NEAR_GAS
+        );
+        callStorageDeposit.then(callback).transact();
+    }
+
+    function storageDepositCallback(
+        IEvmErc20 token,
+        string memory recipientId,
+        bool isRecipientStorageDeposit
+    ) external onlyRole(CALLBACK_ROLE) {
+        require(
+            AuroraSdk.promiseResult(0).status == PromiseResultStatus.Successful,
+            "ERROR: The `storage_deposit()` XCC call failed"
+        );
+
+        if (isRecipientStorageDeposit) {
+            registeredRecipients[token][recipientId] = true;
+            emit RecipientStorageRegistered(token, recipientId);
+        } else {
+            TokenInfo storage tokenInfo = registeredTokens[token];
+            tokenInfo.isStorageRegistered = true;
+            emit TokenStorageRegistered(token, tokenInfo.nearTokenAccountId);
+        }
+    }
+
+    function ftTransferCallToNear(
+        IEvmErc20 token,
+        uint128 amount,
+        string calldata receiverId,
+        string calldata message
+    ) public whenNotPaused {
+        require(near.wNEAR.balanceOf(address(this)) >= ONE_YOCTO, "Not enough wNEAR balance");
+        TokenInfo memory tokenInfo = registeredTokens[token];
+        require(tokenInfo.isStorageRegistered, "The token storage is not registered");
+
+        token.transferFrom(msg.sender, address(this), amount);
+        token.withdrawToNear(bytes(getImplicitNearAccountIdForSelf()), amount);
+        _ftTransferCallToNear(token, tokenInfo.nearTokenAccountId, amount, receiverId, message);
+    }
+
+    function safeFtTransferCallToNear(
+        IEvmErc20 token,
+        uint128 amount,
+        string calldata receiverId,
+        string calldata message
+    ) external whenNotPaused {
+        require(registeredRecipients[token][receiverId], "The storage is not registered for the recipient");
+        ftTransferCallToNear(token, amount, receiverId, message);
+    }
+
+    function _ftTransferCallToNear(
+        IEvmErc20 token,
+        string memory nearTokenAccountId,
+        uint128 amount,
+        string memory receiverId,
+        string memory message
+    ) private {
+        ftTransferCallCounter += 1;
+        emit InitFtTransferCall(ftTransferCallCounter, msg.sender, token, receiverId);
```

Comment:
Do we need to save also amount and message?

Reply:
- I was going to say we can always get the amount from the Transfer event emitted in the same transaction but that wouldn't work with native ETH transfers (will this function be payable to add ETH support ?).
- The use case I had in mind for the `InitFtTransferCall` event is to be able to find the user's original transaction from `FtTransferCall` using the `nonce`. So imo it's ok not to have more parameters.